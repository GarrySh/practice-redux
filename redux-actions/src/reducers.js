import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from './actions';

// BEGIN (write your solution here)
const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    return { ...state, [task.id]: task };
  },
  [actions.removeTask](state, { payload: { id } }) {
    return _.omit(state, id);
  },
  [actions.updateTask](state, { payload: { task } }) {
    return { ...state, [task.id]: task };
  },
}, {});

export default combineReducers({
  tasks,
});
// END
