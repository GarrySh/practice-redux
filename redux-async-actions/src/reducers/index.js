import { omit, keyBy } from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import { reducer as formReducer } from 'redux-form';
import * as actions from '../actions';

// BEGIN (write your solution here)
const taskCreatingState = handleActions({
  [actions.addTaskRequest]() {
    return 'requested';
  },
  [actions.addTaskFailure]() {
    return 'failed';
  },
  [actions.addTaskSuccess]() {
    return 'successed';
  },
}, 'none');
// END

const taskRemovingState = handleActions({
  [actions.removeTaskRequest]() {
    console.log('remove task requested');
    return 'requested';
  },
  [actions.removeTaskFailure]() {
    console.log('remove task failed');
    return 'failed';
  },
  [actions.removeTaskSuccess]() {
    console.log('remove task success');
    return 'successed';
  },
}, 'none');

const tasksFetchingState = handleActions({
  [actions.fetchTasksRequest]() {
    return 'requested';
  },
  [actions.fetchTasksFailure]() {
    return 'failed';
  },
  [actions.fetchTasksSuccess]() {
    return 'successed';
  },
}, 'none');

const tasks = handleActions({
  [actions.fetchTasksSuccess](state, { payload }) {
    return keyBy(payload.tasks, 'id');
  },
  [actions.addTaskSuccess](state, { payload: { task } }) {
    console.log('add task success handle', task);
    return { ...state, [task.id]: task };
  },
  [actions.removeTaskSuccess](state, { payload: { task } }) {
    console.log('remove task success handle', task);
    return omit(state, task.id);
  },
  [actions.toggleTaskState](state, { payload: { id } }) {
    const task = state[id];
    const newState = task.state === 'active' ? 'finished' : 'active';
    const updatedTask = { ...task, state: newState };
    return { ...state, [task.id]: updatedTask };
  },
}, {});

export default combineReducers({
  taskCreatingState,
  taskRemovingState,
  tasksFetchingState,
  tasks,
  form: formReducer,
});
