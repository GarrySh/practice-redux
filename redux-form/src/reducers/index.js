import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import { reducer as formReducer } from 'redux-form';
import * as actions from '../actions';

const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    return { ...state, [task.id]: task };
  },
  [actions.removeTask](state, { payload: { id } }) {
    return _.omit(state, id);
  },
  [actions.toggleTaskState](state, { payload: { id } }) {
    const task = state[id];
    const newState = task.state === 'active' ? 'finished' : 'active';
    const updatedTask = { ...task, state: newState };
    return { ...state, [task.id]: updatedTask };
  },
}, {});

export default combineReducers({
  // BEGIN (write your solution here)
  form: formReducer,
  // END
  tasks,
});
