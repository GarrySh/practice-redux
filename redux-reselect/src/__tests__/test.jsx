import 'js-polyfills/html';

import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import reducers from '../reducers';
import App from '../components/App';

Enzyme.configure({ adapter: new Adapter() });

test('Store', () => {
  const store = createStore(reducers);

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );
  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('input[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'na-na' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const links = wrapper.find('.app-remove-task');
  links.last().simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const link = wrapper.find('.app-toggle-state');
  link.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  link.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
