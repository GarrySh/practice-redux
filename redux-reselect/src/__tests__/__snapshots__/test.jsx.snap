// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
</div>
`;

exports[`Store 2`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="na-na"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
</div>
`;

exports[`Store 3`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          na-na
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="another task"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          na-na
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          na-na
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          another task
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          na-na
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 7`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          <s>
            na-na
          </s>
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 8`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex justify-content-end"
      >
        <button
          class="btn border-0 p-0 app-toggle-state mr-3"
        >
          -
        </button>
        <div
          class="mr-auto"
        >
          na-na
        </div>
        <button
          class="btn border-0 p-0 app-remove-task"
        >
          x
        </button>
      </li>
    </ul>
  </div>
</div>
`;
